### PR TITLE
0.3.0 API

### PR DESCRIPTION
**Pull request by jballanc**
Updated API for v0.3.0 that allows for efficient single-call decode of byte arrays or files to either a pre-allocated `ByteBuffer` or returned as a byte array.

### CODE REVIEW

**Comment by chris-allan** [comment]
@melissalinkert should have a look at see what she thinks of the new API. Is there a particular reason the ByteBuffer tests were disabled in dcf512a, @jballanc?

**Comment by jballanc** [comment]
Yes, `ByteBuffer` tests were failing due to the issues we saw before with creating too many `ByteBuffer`s. I suspect this may have to do with the changes to `getImageMetadata`, i.e. there's twice the number of `ByteBuffer`s in play because we're no longer assigning the `ByteBuffer` to an instance and re-using it, but rather passing it as a parameter.

One option would be to change the API specifically for `ByteBuffer`s such that we can instantiate `DecodeContext` with a `ByteBuffer` and re-use it, but then the API for `ByteBuffer` would be different than for `File` or `byte[]` which seems less than idea.

I asked @melissalinkert if support for `ByteBuffer`s was critical, and she indicated it was not, so I had intended to simply remove those bits and simplify the API. Working on a proposal for than in a branch of this branch.

File: java/src/test/java/ome/jxrlib/TestInMemoryDecode.java [hunk 1]
```diff
@@ -99,27 +131,31 @@ public void testByteArray(
...
     @Test
     public void testFirstFrame(String filename, String md5)
             throws IOException, DecodeException {
         byte[] data = asByteArray(filename);
-        byte[] destination = TestDecode.decodeFirstFrame(data, 0, data.length);
+        byte[] destination = new TestDecode().decodeFrame(0, data);
         Assert.assertEquals(md5, md5(destination));
     }
 
+    /*
```
Comment (chris-allan) [inline java/src/test/java/ome/jxrlib/TestInMemoryDecode.java]: Let's add a TestNG annotation here and comment as disabled for now or adjust the test cases to re-use a ByteBuffer if it's a memory issue we're running into.

File: java/src/test/resources/testng.xml [hunk 2]
```diff
@@ -3,6 +3,19 @@
 
 <suite name="ome.jxrlib">
 
+  <test name="FL_3channel_noZ_compression_lossless">
```
Comment (chris-allan) [inline java/src/test/resources/testng.xml:6]: Is there a particular reason this was moved up?

**Comment by chris-allan** [comment]
I can't see a reason not to just make it work by adjusting the test cases. We know the size of the required buffers as well based on the test case fixture data being passed in and we can pre-allocate, pass in and/or reuse the buffer as required by setting the TestNG scope.

File: java/src/test/resources/testng.xml (hunk 2 above)
Comment (jballanc) [reply java/src/test/resources/testng.xml:6]: It's the largest of the test files, so moving it was moved up in the hope that (combined with the buffer reallocation workaround in e29028c) running it first would avoid the issues with `ByteBuffer`s. (This worked on my machine, but apparently not on Travis.)

**Comment by jballanc** [comment]
We could certainly re-jigger the tests to avoid excessive `ByteBuffer` allocation. I already attempted a janky solution along these lines in e29028c, but apparently to no avail. If keeping the `ByteBuffer` APIs around is a priority, I can look at this again, but my impression was that those APIs were work-arounds until direct decoding of `byte[]` was possible (which it is now), and so they aren't really needed any longer.

Or, rather, it's not necessary to use `ByteBuffer`s with `DecodeContext`. We could certainly keep the `ByteBuffer` API in `ImageDecode` and `CodecFactory`, if desired, for those that want to separate `ImageDecode` creation and image decoding. In that case, though, we should write tests that exercise those APIs directly (and avoid more `ByteBuffer`s than necessary).

**Comment by chris-allan** [comment]
They're not really a work around but rather an alternative.

The first thing I would try to allocate the buffers ahead of time in the test case itself and utilise TestNG scoping to ensure that they are not destroyed.

**Comment by chris-allan** [comment]
Two things, @jballanc:

 * Lets get some test cases added for decoding into a larger `ByteBuffer` at a given offset
 * We should probably have `getImageMetadata(ByteBuffer source)` added to the API and tests added for it

**Comment by melissalinkert** [comment]
Java API looks good in general.  I could see it being useful to have something like ```AbstractDecode.decodeFrame(int frame, byte[] source, int sourceOffset, int sourceLength, byte[] destination, int destinationOffset)```, especially for consistency with the other ```decodeFrame``` signatures that take a ```File``` or ```ByteBuffer```.  It might also be helpful to have place-holder ```getFrameCount``` methods that just returns 1 - we know that there is only one frame in practice, but that won't be obvious to anyone external.  Adding a frame count field to ```ImageMetadata``` would also work, if that's easier.

Otherwise, I think it would be good to do a little more bounds checking in the ```decodeFrame``` methods, to ensure that the frame and offset are non-negative and the length is positive.

**Comment by melissalinkert** [comment]
Closing as discussed in the context of upcoming work on #32.

**Pull request closed without merging**